kernel/src/time/wait.rs

```diff
@@ -5,48 +5,109 @@ use core::time::Duration;
+/// A trait that provide the timeout related function for [`Waiter`] and [`WaitQueue`]`.
 pub trait WaitTimeout {
     /// Wait until some condition returns `Some(_)`, or a given timeout is reached. If
-    /// the condition does not becomes `Some(_)` before the timeout is reached, the
-    /// function will return `None`.
-    fn wait_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Option<R>
+    /// the condition does not becomes `Some(_)` before the timeout is reached or `cancel_cond`
+    /// returns `Err`, this function will return corresponding `Err`.
+    fn wait_until_or_timeout<F, R, FCancel>(
+        &self,
+        cond: F,
+        cancel_cond: FCancel,
+        timeout: &Duration,
+    ) -> Result<R>
     where
-        F: FnMut() -> Option<R>;
+        F: FnMut() -> Option<R>,
+        FCancel: Fn() -> Result<()>;
 }
 
-impl WaitTimeout for WaitQueue {
-    fn wait_until_or_timeout<F, R>(&self, mut cond: F, timeout: &Duration) -> Option<R>
+impl WaitTimeout for Waiter {
+    fn wait_until_or_timeout<F, R, FCancel>(
+        &self,
+        mut cond: F,
+        cancel_cond: FCancel,
+        timeout: &Duration,
+    ) -> Result<R>
     where
         F: FnMut() -> Option<R>,
+        FCancel: Fn() -> Result<()>,
     {
         if *timeout == Duration::ZERO {
-            return cond();
+            return cond()
+                .ok_or_else(|| Error::with_message(Errno::ETIME, "the time limit is reached"));
         }
 
         if let Some(res) = cond() {
-            return Some(res);
+            return Ok(res);
         }
 
-        let (waiter, waker) = Waiter::new_pair();
-
+        let waker = self.waker();
         let jiffies_timer = JIFFIES_TIMER_MANAGER.get().unwrap().create_timer(move || {
             waker.wake_up();
         });
         jiffies_timer.set_timeout(Timeout::After(*timeout));
 
-        let cancel_cond = {
+        let timeout_cond = {
             let jiffies_timer = jiffies_timer.clone();
-            move || jiffies_timer.remain() == Duration::ZERO
+            move || {
+                if jiffies_timer.remain() != Duration::ZERO {
+                    Ok(())
+                } else {
+                    Err(Error::with_message(
+                        Errno::ETIME,
+                        "the time limit is reached",
+                    ))
+                }
+            }
+        };
+
+        let cancel_cond = || {
+            timeout_cond()?;
+            cancel_cond()?;
+            Ok(())
         };
-        let res = self.wait_until_or_cancelled(cond, waiter, cancel_cond);
+
+        let res = self.wait_until_or_cancelled(cond, cancel_cond);
 
         // If res is `Some`, then the timeout may not have been expired. We cancel it manually.
-        if res.is_some() {
+        if !res
+            .as_ref()
+            .is_err_and(|e: &Error| e.error() == Errno::ETIME)
+        {
             jiffies_timer.cancel();
```

Comment:
If res is Err but errno is not `ETIME`, the timer should also be cancelled, right?